### DISCUSSION TITLE
How to Install Custom dts/dtbo?

### DISCUSSION THREAD

**Discussion by hftsai256** [nix-community/raspberry-pi-nix#77] [Q&A] [answered]
I'm seeking a way to build a Raspberry Pi CM4 image with several customized dts/dtbo, e.g. [mcp4728.dts](https://gist.github.com/hftsai256/6e8c3111649b48d06df6e703caaaa71d#file-mcp4728-dts). My first intuition is something like this:
```
hardware.deviceTree = {
  kernelPackage = pkgs.linux_rpi4;
  filter = "bcm2711-rpi-cm4.dtb";
  overlays = [
    { name = "mcp4728"; dtsFile = ./mcp4728.dts; }
    { name = "spi6-1cs"; dtboFile = "${pkgs.device-tree_rpi.overlays}/spi6-1cs.dtbo"; }
  ];
};
```
But then I find `boot.loader.generic-extlinux-compatible.useGenerationDeviceTree = false`, and I doubt the path I'm taking will be effective.

My another thought is to create a separate Nix package to handle the build/install part:
```
{ stdenvNoCC, dtc }:
stdenvNoCC.mkDerivation {
  pname = "dtoverlays";
  version = "0.0.1";
  src = ./.;
  buildInputs = [ dtc ];
  installPhase = ''
    install -d $out/boot/firmware/overlays
    mv build/*.dtbo $out/boot/firmware/overlays
  '';
}
```
However as far as I concerned, the install path `/boot/firmware/overlays` is in a different partition from `rootfs`, and therefore I also doubt that I can include this package in `environment.systemPackages`

 Is there a way to build the dtbo from source and install them into `/boot/firmware/overlays`?

**Comment by tstat** [comment]
Hey @hftsai256. The short answer to your question is that overlays are taken from the package `raspberrypifw`, so you could override that to add an overlay. This is mentioned in the readme under [this section](https://github.com/nix-community/raspberry-pi-nix?tab=readme-ov-file#firmware-partition-implementation-notes).

Reading the firmware populating script might clarify how this all fits together in your mind. Specifically, the device tree overlays are populated [here](https://github.com/nix-community/raspberry-pi-nix/blob/master/rpi/default.nix#L159-L181).

**Comment by hftsai256** [comment]
Thank you for your directions. I'm wondering if I can have a separate package that installs into the same path: in this case, `$out/share/raspberrypi/boot/firmware/overlays`, before `migrate_firmware()` is invoked.

I have set up the following structure that may be viewed as a standalone nix package:
```
dtoverlays
├── default.nix
├── flake.nix
├── Makefile
└── mcp4728.dts
```

On the other hand, I would try giving overriding `raspberrypifw` a shot. My initial thought would be something like this:
```
  outputs = { self, nixpkgs, nixos-hardware } @ inputs: let
    pkgs = import nixpkgs {
      system = "aarch64-linux";
      overlays = [ customDtOverlay ];
    };
    customDtOverlay = (final: prev: let
      newfw = prev.raspberrypifw.overrideAttrs (oldAttrs: {
        srcs = [ oldAttrs.src ./dtoverlays ];
        sourceRoot = ".";
        buildInputs = [ pkgs.dtc ];

        dontBuild = false;
        buildPhase = ''
          make -C ./dtoverlays
        '';

        installPhase = ''
          mkdir -p $out/share/raspberrypi
          mv ./source/boot $out/share/raspberrypi
          mkdir -p $out/share/raspberrypi/boot/overlays
          mv ./dtoverlays/build/*.dtbo $out/share/raspberrypi/boot/overlays
        '';
      });
  in {
  ...
  }
```
I have to override the entire `installPhase` as the paths are changed after modifying srcs

**Reply by tstat** [reply, marked as answer]
I'm not against the idea of having a list of packages that contain overlays.

An alternative to overriding the derivation with an additional source as you demonstrated here might to override `raspberrypifw` with the `symlinkJoin` of your package and the prev `raspberrypifw`.

**Reply by hftsai256** [reply]
Thanks for the help. I can sort of building my customized `raspberypifw` with the overlay above. However the overridden package would not apply to the SD image unless `pin-inputs.enable = false` (as I came across with #83 and realize this could be the key to my last puzzle.) The down side is to rebuild the entire Linux kernel under qemu, but I'm having a working image now.

I'm now on the journey of trimming the image down to ~500-800 MB instead of the current whopping 6 GB, but that's a different topic.
